File: 1774.closest-dessert-cost.cs
/*
 * @lc app=leetcode id=1774 lang=csharp
 *
 * [1774] Closest Dessert Cost
 */

// @lc code=start
using System;

public class Solution {
    private int[] baseCosts;
    private int[] toppingCosts;
    private int target;
    private int result;
    private int resultDiff;

    public int ClosestCost(int[] baseCosts, int[] toppingCosts, int target) {
        this.baseCosts = baseCosts;
        this.toppingCosts = toppingCosts;
        this.target = target;
        this.result = -1;
        this.resultDiff = int.MaxValue;

        for (int i = 0; i < baseCosts.Length; i++) {
            Backtracking(0, baseCosts[i] - target);
        }

        return result;
    }

    private void Backtracking(int i, int diff) {
        if ((diff > 0 && Math.Abs(diff) < resultDiff) || (diff <= 0 && Math.Abs(diff) <= resultDiff)) {
            result = target + diff;
            resultDiff = Math.Abs(diff);
        }

        if (diff < 0 && i < toppingCosts.Length) {
            for (int j = 0; j < 3; j++) {
                Backtracking(i + 1, diff + j * toppingCosts[i]);
            }
        }
    }
}

// @lc code=end

File: 1774.closest-dessert-cost.cpp
/*
 * @lc app=leetcode id=1774 lang=cpp
 *
 * [1774] Closest Dessert Cost
 */

// @lc code=start
#include<vector>
#include<algorithm>
using std::abs;
using std::vector;
class Solution {
private:
    vector<int> ToppingCosts;
    int Target;
    int result = -1;
    int resultDiff = INT_MAX;
    void Backtracking(int i, int diff) {
        if ((diff > 0 && abs(diff) < resultDiff) || (diff <= 0 && abs(diff) <= resultDiff)) {
            result = Target + diff;
            resultDiff = abs(diff);
        }

        if (diff < 0 && i < ToppingCosts.size()) {
            for (int j = 0; j < 3; j++) {
                Backtracking(i + 1, diff + j * ToppingCosts[i]);
            }
        }
    }

public:
    int closestCost(vector<int>& baseCosts, vector<int>& toppingCosts, int target) {
        Target = target;
        ToppingCosts = toppingCosts;

        for (int i = 0; i < baseCosts.size(); i++) {
            Backtracking(0, baseCosts[i] - target);
        }

        return result;
    }
};
// @lc code=end

